ddtrace/contrib/datadog_lambda/patch.py

```diff
@@ -0,0 +1,148 @@
+from importlib import import_module
+import os
+import signal
+
+from ddtrace import tracer
+from ddtrace.constants import ERROR_MSG
+from ddtrace.constants import ERROR_TYPE
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.internal.wrapping import unwrap
+from ddtrace.internal.wrapping import wrap
+
+
+log = get_logger(__name__)
+
+
+def crash_flush(_, __):
+    """
+    Tags the current root span with an Impending Timeout error and
+    finishes all remaining spans.
+
+    Shutdowns the tracer at the end.
+    """
+
+    root_span = tracer.current_root_span()
+    root_span.error = 1
+    root_span.set_tag_str(ERROR_MSG, "Datadog detected an Impending Timeout")
+    root_span.set_tag_str(ERROR_TYPE, "Impending Timeout")
+
+    current_span = tracer.current_span()
+    while current_span:
+        current_span.finish()
+        current_span = current_span._parent
```

Comment:
This internal usage isn't good as it'll lock us into the underlying implementation details and prevent us from making easier refactors in the future.

I do see that there isn't a public API available that provides this functionality. I think it's worth introducing one for this use-case. How does https://github.com/DataDog/dd-trace-py/compare/span-finish-parents look?

Reply:
Yes, that makes sense, it looks good!
Do you want me to go ahead and create the PR based on your commits?

Reply:
Opened here: https://github.com/DataDog/dd-trace-py/pull/5093